src/main/java/com/ibm/cloudant/kafka/connect/CachedClientManager.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright © 2022 IBM Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+package com.ibm.cloudant.kafka.connect;
+
+import com.ibm.cloud.cloudant.internal.ServiceFactory;
+import com.ibm.cloud.cloudant.v1.Cloudant;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.ibm.cloudant.kafka.common.utils.JavaCloudantUtil.VERSION;
+
+public class CachedClientManager {
+
+    // Cloudant clients, keyed by connector name
+    static Map<String, Cloudant> clientCache = new ConcurrentHashMap<>();
+
+    private CachedClientManager() {
+        // no instantiation
+    }
+
+    public static Cloudant getInstance(Map<String, String> props) {
+        String connectorName = props.get("name");
+        return clientCache.computeIfAbsent(connectorName, p -> ServiceFactory.getInstance(props, VERSION));
+    }
+
+    public static void stop(Map<String, String> props) {
+        String connectorName = props.get("name");
+        clientCache.remove(connectorName);
```

Comment:
Have you observed any delayed termination of e.g the test workers? Do you think it is worth shutting down the underlying client here with something like:

```suggestion
        Optional.ofNullable(clientCache.remove(connectorName)).ifPresent(cloudant -> {
            OkHttpClient finishedClient = cloudant.getClient();
            finishedClient.dispatcher().executorService().shutdown();
            finishedClient.connectionPool().evictAll();
        });
```

[OkHttp suggests that shutdown isn't necessary](https://square.github.io/okhttp/4.x/okhttp/okhttp3/-ok-http-client/#shutdown-isnt-necessary) as idle clients cleanup eventually, but I have seen the ok thread keep things alive for _O(minutes)_ after stopping work. Not sure how impactful that is, I guess in a standalone mode it could be annoying, but in distributed Kafka it probably isn't important. I think the downside is interacting with the underlying client implementation so we should probably only do it if we've seen issues.

Reply:
That's an interesting point, but I think I'm happy with OkHttp's advice here, because:

- In a production environment it's unlikely that users will delete and create connectors rapidly or often via the REST interface
- It's actually beneficial to leave a connection idle in the pool. because if a connector is re-created with the same Cloudant host within the timeout period, then that connection can be re-used.